Add a builtin ML implementation

ggeorgakoudis opened the pull request:
- Add implementation of decision tree, random forest
- Make OpenCV an optional dependency (ENABLE_OPENCV)
- Add env var (APOLLO_TUNING_MODEL) to select DecisionTree or RandomForest models
- Add python scripts to visualize with models with dot
- Update Kokkos connector

DavidPoliakoff commented on src/connectors/kokkos/kokkos-connector.cpp:
```diff
@@ -491,7 +492,7 @@ extern "C" void kokkosp_request_values(
       tuning_regions.emplace(tuningProblem, std::make_pair("", nullptr));
   if (iter.second) {
-    std::string prefix = "dtree-step-0-rank-0-";
+    std::string prefix = "RandomForest-step-0-rank-0-";
```
Lemme test these locally, I'll let you know if it works out

ggeorgakoudis replied:
Thanks! I was trying to add you as a reviewer but your name doesn't appear in the list. Any idea why?

davidbeckingsale commented on src/Apollo.cpp:
```diff
@@ -544,102 +518,115 @@ Apollo::train(int step) {
                   ModelFactory::createRegressionTree(train_time_features,
                                                      train_time_responses);
+#else
+              assert(false && "Retraining requires OpenCV");
```
This assert could be compiled out, so mabye throw an exception instead?

ggeorgakoudis replied:
I agree that handling is not ideal. The ifdef-conditional assert is intentional. We do not (yet) support regression in our built-in ML library. The assertion will abort execution if the user enables re-training through the env var but has not compiled with ENABLE_OPENCV. Could we accept as a stopgap until regression falls in?

ggeorgakoudis replied:
Fixed through exceptions

cdwdirect commented:
Yes. I agree about implementing an internal DT.

> On Nov 2, 2021, at 8:09 PM, Giorgis Georgakoudis ***@***.***> wrote:
> 
> ﻿
> @ggeorgakoudis commented on this pull request.
> 
> In src/Apollo.cpp:
> 
> >                reg->time_model =
>                    ModelFactory::createRegressionTree(train_time_features,
>                                                       train_time_responses);
> +#else
> +              assert(false && "Retraining requires OpenCV");
> I agree that handling is not ideal. The ifdef-conditional assert is intentional. We do not (yet) support regression in our built-in ML library. The assertion will abort execution if the user enables re-training through the env var but has not compiled with ENABLE_OPENCV. Could we accept as a stopgap until regression falls in?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

ggeorgakoudis committed:
Add a builtin ML implementation

- Add implementation of decision tree, random forest
- Make OpenCV an optional dependency (ENABLE_OPENCV)
- Add env var (APOLLO_TUNING_MODEL) to select DecisionTree or RandomForest models
- Add python scripts to visualize with models with dot
- Update Kokkos connector
